### ISSUE TITLE
Automatic differentiation to calculate Jacobians

### ISSUE THREAD

**Issue by mikeingold** [JuliaGeometry/MeshIntegrals.jl#35]
The codebase currently uses an inelegant but reasonable finite difference approximation to calculate the Jacobian of a geometry's parametric function.

https://github.com/mikeingold/MeshIntegrals.jl/blob/66ed919150b3774bf09dabab60476869c5514493/src/utils.jl#L5-L15

At time of writing, there didn't appear to be an automatic differentiation package that could compute the same thing out-of-the-box. These parametric functions return a `Point` rather than a simple number. I believe this could be addressed by defining chain rules in Meshes.jl, but I'm open to other solutions.

**Comment by mikeingold** [comment]
Meshes.jl now uses Zygote.jl as an AD-of-choice. Ideally, all `Geometry`s in Meshes.jl would have diff rules established so that this Jacobian function would flow from that.

**Comment by mikeingold** [comment]
Merged with #37

**Comment by mikeingold** [comment]
Since `differential` was generalized in #101, I'm going to un-merge this Issue from #37.

**Comment by JoshuaLampert** [comment]
In #115 we agreed that it would be best to implement the AD support over in Meshes.jl (we could also think about transferring the issue?).
I haven't tested anything AD related in the context of Meshes.jl and MeshIntegrals.jl (I only have a bit of experience using Zygote.jl, ForwardDiff.jl and Enzyme.jl in other contexts), but I see that the fact that the parametric functions of the geometries return `Point`s complicates the issue and probably necessitates implementing custom AD rules via ChainRules.jl.
> I’m also not sure if writing these directly for parametric functions is the best path, or if we can write them for lower-level functions and have that support roll up automatically.

I would opt for minimizing the number of rules to implement, i.e. not needing to implement a rule for each geometry, but rather write something more low level. This seems to be easier to maintain and extend.
These are just my initial thought without doing any experimenting myself. My familiarity with ChainRules.jl is pretty low. So I'm not sure if this approach is feasible.
Getting some advice from @kylebeggs would be awesome.

**Comment by mikeingold** [comment]
> ... but I see that the fact that the parametric functions of the geometries return `Point`s complicates the issue and probably necessitates implementing custom AD rules via ChainRules.jl.

Yes, I'm definitely not an AD expert but this is one of the reasons that my intuition tells me we'll need custom AD rules. Parametric functions map from floating point tuples/vectors to `Meshes.Point`, but their derivatives should in my opinion map from floating point tuples/vectors to `Meshes.Vec`.

Again, though, I'm not sure if there's a simpler way to define rules like the derivative of `Point(...)` is `Vec(...)`, etc and get those to propagate automatically. This would be especially beneficial in that it wouldn't require AD rule updates any time the corresponding functions are updated.

**Comment by kylebeggs** [comment]
I highly encourage rules to be implemented at the lowest level possible. Usually, once you have those, everything else will "just work" and then the only reason to write more rules is for efficiency (i.e. you know some computations in the graph for autodiff can be skipped or something via knowledge of the math behind the operation). I think for Meshes.jl, the hardest part will really be defining what it means for some of these types/methods to be differentiated, as @mikeingold alluded to.

Although the rules should go in Meshes.jl, we can use MeshIntegrals.jl as a catalyst for figuring out what rules even need to be defined to begin with. We can create a matrix of to-dos like there was in the early stages of the package in the README.

**Comment by kylebeggs** [comment]
I suppose that would be just any types `<:Geometry`

**Comment by mikeingold** [comment]
> We can create a matrix of to-dos like there was in the early stages of the package in the README.

For reference, the Support Matrix you’re probably referring to is now hosted on our Documenter site.

**Comment by mikeingold** [comment]
For anyone experimenting with AD on parametric functions, I’d probably recommend starting experiments with `Box` in 1-D space, then 2-D, etc. In 1-D it’s basically just an interval/segment on a number line, but the parametric function is still generic enough to handle multiple dimensions/arguments.

**Comment by kylebeggs** [comment]
I started with `Segment`, but you make a good point I'll switch to `Box`. Another thing to consider here is Zygote/ChainRules vs Enzyme. Do we want to support both? Or just pick one? I find Enzyme to be faster, and until recently there wasn't an option to write your own rules. I have not had a chance to use the new-ish rule system in Enzyme. Enzyme works out of the box for `Segment`, but returns a `Point` so we need a rule just because I think we established already it should be returning `Vec`?

**Comment by kylebeggs** [comment]
> but I'd be curious
to see if something more complex like the jacobian of a 3D geometry could
be made to work without needing custom rules.

I will try this out and if it seems to work, we can add support directly to MeshIntegrals.jl while we continue to plan AD support in Meshes.jl.

@juliohm can you comment on supporting Zygote and/or Enzyme?

**Comment by juliohm** [comment]
Hi @kylebeggs I don't have a strong preference here, but I am not the most qualified person to answer.

We adopted Zygote.jl in CoordRefSystems.jl because it was there for a long time and it is native Julia. I tend to go with native Julia deps as much as possible to avoid situations where the host platform is supported by Julia, but the binary dependency is not.

Adopting Zygote.jl could be interesting in the sense that we already depend on it through CoordRefSystems.jl

**Comment by kylebeggs** [comment]
Given the code in Meshes.jl is pretty functional (in style), we can begin with Zygote. I find for code that depends on heavy pre-allocations it is a must to use Enzyme.

Enzyme seems to work directly for >1D as well:
```julia-repl
julia> b=Box((0,0),(1,1))
Box
├─ min: Point(x: 0.0 m, y: 0.0 m)
└─ max: Point(x: 1.0 m, y: 1.0 m)

julia> to.(Enzyme.jacobian(Forward, b, 0.5, 0.5))
(Vec(1.0 m, 0.0 m), Vec(0.0 m, 1.0 m))
```

**Comment by mikeingold** [comment]
> Enzyme seems to work directly for >1D as well:

That example is genuinely very exciting for me to see. In the interim until Zygote chain rules are implemented in Meshes, I would very seriously consider rewriting our `jacobian` to short-cut to this internally. I'm also very curious what kind of performance difference this would produce vs finite-diff, so good thing I just wrapped up #119.

**Comment by mikeingold** [comment]
@kylebeggs, would you be up for writing a PR that adds Enzyme as a dependency and switches to using a Jacobian like you just showed? Maybe making our `Jacobian` something like a one-liner wrapper like that in example?

**Comment by JoshuaLampert** [comment]
This looks already really great. Thanks a lot for your effort @kylebeggs!

**Comment by kylebeggs** [comment]
I also got Zygote working for `Segment` with ChainRules defined, but I think this is just 2 different issues. I'll post up an example later, perhaps make a small PR to Meshes.jl to get the ball rolling. I'll do this one first, though

**Comment by mikeingold** [comment]
Should we expect a Meshes.jl PR to be approved soon that enables AD-able parametric functions via Zygote chain rules for the same kind of `Jacobian` result? If so, it might be best to work on that first and then see if it “just works” from here without Enzyme. If it looks like that process could take significantly longer, though, then I’d probably just go with Enzyme here in the interim.

@JoshuaLampert, I was just thinking: if chain rules are added to the next minor version release of Meshes and we centrally rely on those for calculating integrals, then we might need to bump our minimum dep version accordingly. Enzyme would obviously be backwards compatible, though. I suppose there also might be another route where we somehow conditionally define `jacobian` to point toward Meshes chain rules or Enzyme depending on version. I know you’ve mentioned a desire to maintain compat with older versions, so how do you feel about this?

**Comment by JoshuaLampert** [comment]
We could also go another route to allow compatibility, avoid breaking changes, and also have maximal flexibility: We could add a kwarg to `integral` (and `lineintegral` etc.) `autodiff` (or whatever) defaulting to `:finite_difference`, which calls the current FD implementation. We could then add the option `:zygote`, which calls the Zygote `jacobian` that will be added to Meshes.jl. This could then be allowed conditionally on the version of Meshes.jl. If adding Enzyme.jl support is really as straightforward as @kylebeggs mentioned above, we could also think of additionally adding the `option` `:enzyme`. Optim.jl does something similar for their diff backend.

**Comment by mikeingold** [comment]
For specifying the type of differentiation method, I was looking to see how these are specified in SciML’s Optimization.jl. [Their docs outline](https://docs.sciml.ai/Optimization/stable/API/ad/nothing/Optimization/stable/API/ad/) how they use the [SciML/ADTypes.jl package](https://github.com/SciML/ADTypes.jl) for this. Using those directly could be interesting but would definitely add some complexity.

**Comment by kylebeggs** [comment]
I think that package (or a similar approach) should be the long term goal, but I don't want to set the bar too high for the initial foray into AD unless using this package isn't much more involved to implement but I haven't used it or explored it yet.

**Comment by mikeingold** [comment]
They define `AutoZygote` and `AutoEnzyme` singletons that would be easy enough to treat as dispatch flags, but their finite diff type is defined as falling back on FiniteDiff.jl, so we`d either need to additionally rework our finite diff implementation or specify that we break from that. We’d also potentially need to either implement all the other types they define or not-implemented errors.

Overall, I don’t think there’s a particularly strong reason to use that package, especially given the overhead, but I would consider using a consistent naming scheme for types and kwargs.

**Comment by JoshuaLampert** [comment]
I agree. I don't really see the gain in using ADTypes.jl. Defining our own structs to dispatch on seems like a good solution.
I also agree on a consistent naming scheme.

**Comment by kylebeggs** [comment]
So all tests pass, _except_ for `Cylinder`, `CylinderSurface`, and `ParameterizedCurve`. I will try to debug those while the new API in #122 gets finalized (looks good!), then I'll conform to that and make a PR.

**Comment by mikeingold** [comment]
> So all tests pass, _except_ for `Cylinder`, `CylinderSurface`, and `ParameterizedCurve`. I will try to debug those while the new API in #122 gets finalized (looks good!), then I'll conform to that and make a PR.

Do the tests fail or error? The [parametric function for `Cylinder`](https://github.com/JuliaGeometry/Meshes.jl/blob/b173cf34b4ad99413545633670688aad5395b85d/src/geometries/primitives/cylinder.jl#L74-L99) seems to have gotten surprisingly complex. The parametric function for `CylinderSurface` just treats it as a degenerate case of `Cylinder(rho, phi, z)` where $\rho=1$.

**Comment by kylebeggs** [comment]
Sorry, they error.

**Comment by mikeingold** [comment]
@JoshuaLampert I think we should mark this issue as closed by #152. For remaining future work, like Zygote, I think we should have new issues with narrower scope (better defined achievables and closing criteria).

**Comment by JoshuaLampert** [comment]
I agree.
